matrix_client/room.py

```diff
@@ -477,23 +474,21 @@ def add_room_alias(self, room_alias):
     def get_joined_members(self):
         """Returns list of joined members (User objects)."""
         if self._members:
-            return self._members
+            return list(self._members.values())
         response = self.client.api.get_room_members(self.room_id)
         for event in response["chunk"]:
             if event["content"]["membership"] == "join":
-                self._mkmembers(
-                    User(self.client.api,
-                         event["state_key"],
-                         event["content"].get("displayname"))
-                )
-        return self._members
-
-    def _mkmembers(self, member):
-        if member.user_id not in [x.user_id for x in self._members]:
-            self._members.append(member)
-
-    def _rmmembers(self, user_id):
-        self._members[:] = [x for x in self._members if x.user_id != user_id]
+                self._add_member(event["state_key"], event["content"].get("displayname"))
```

Comment:
Yeah. I doesn't look like we have access to the global displayname here though.  
So the fix would be to maintain a map from room IDs to displaynames in `User`, and to change the signature of `User.get_displayname(self)` to `User.get_displayname(self, room)`, knowing that the first `user.get_displayname()` will always trigger a request to fetch the global displayname we don't have. Not super nice, what do you think?  
Also the map of displaynames might get used as a hack to know in which rooms a particular user is, so we may want to implement something that allows that cleanly beforehand.

Reply:
Waiting on your approval on this to implement this not so satisfying solution + a new attribute of `User` which contains the `Room`s it is a member of.

Reply:
Sorry on the delay here. I think the map mxid->displayname should live on the room object rather than the user object since that's how it's implemented in the protocol (per-room displaynames are just another state event). This also makes the hack about which rooms a user is in a bit more difficult, so we can kick that can down the road a bit.

Reply:
Done.